File: java/var-core/src/test/java/com/oselvar/var/core/TableCellsTest.java
package com.oselvar.var.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.oselvar.var.core.TableCells.RowCells;
import java.util.List;
import org.junit.jupiter.api.Test;

/**
 * Port of the table-row-cell-parsing behavior of {@code table-cells.ts}, cross-checked against
 * {@code python/packages/var-core/tests/test_table_cells.py} (which in turn ports the table-cell
 * span cases of {@code var-core/tests/scanner.test.ts}).
 */
class TableCellsTest {

    @Test
    void basicRowReturnsTrimmedCells() {
        String source = "| a | b |";
        RowCells result = TableCells.parseRowCells(source, 0, source);
        assertEquals(List.of("a", "b"), result.cells());
    }

    @Test
    void basicRowSpansPointToTrimmedText() {
        String source = "| a | b |";
        RowCells result = TableCells.parseRowCells(source, 0, source);
        assertEquals(2, result.cellSpans().size());
        assertEquals("a", slice(source, result.cellSpans().get(0)));
        assertEquals("b", slice(source, result.cellSpans().get(1)));
    }

    @Test
    void extraPaddingIsTrimmed() {
        String source = "| Bob  | 30  |";
        RowCells result = TableCells.parseRowCells(source, 0, source);
        assertEquals(List.of("Bob", "30"), result.cells());
        assertEquals("Bob", slice(source, result.cellSpans().get(0)));
        assertEquals("30", slice(source, result.cellSpans().get(1)));
    }

    @Test
    void noPipeReturnsEmpty() {
        String source = "hello world";
        RowCells result = TableCells.parseRowCells(source, 0, source);
        assertEquals(List.of(), result.cells());
        assertEquals(List.of(), result.cellSpans());
    }

    @Test
    void singlePipeReturnsEmpty() {
        String source = "| only one";
        RowCells result = TableCells.parseRowCells(source, 0, source);
        assertEquals(List.of(), result.cells());
        assertEquals(List.of(), result.cellSpans());
    }

    @Test
    void singleColumnTableRow() {
        String source = "| n |";
        RowCells result = TableCells.parseRowCells(source, 0, source);
        assertEquals(List.of("n"), result.cells());
        assertEquals("n", slice(source, result.cellSpans().get(0)));
    }

    @Test
    void lineStartOffsetShiftsSpans() {
        String prefix = "# T\n\n";
        String row = "| a | b |";
        String source = prefix + row;
        int lineStart = prefix.length();
        RowCells result = TableCells.parseRowCells(row, lineStart, source);
        assertEquals(List.of("a", "b"), result.cells());
        assertEquals("a", slice(source, result.cellSpans().get(0)));
        assertEquals("b", slice(source, result.cellSpans().get(1)));
    }

    /**
     * A cell containing an astral character (U+1F389, a surrogate pair — 2 UTF-16 code units)
     * must shift the following cell's span by 2, not 1. Java's {@code String}/{@code char} are
     * already UTF-16 code-unit indexed (see {@link Span}'s javadoc), so this needs no conversion
     * layer — unlike the Python port, which reconstructs UTF-16 offsets explicitly.
     */
    @Test
    void astralCellShiftsFollowingSpan() {
        String source = "| 🎉 | a |"; // U+1F389 PARTY POPPER
        RowCells result = TableCells.parseRowCells(source, 0, source);
        assertEquals(List.of("🎉", "a"), result.cells());
        assertEquals(2, result.cellSpans().get(0).startOffset());
        assertEquals(4, result.cellSpans().get(0).endOffset());
        assertEquals(7, result.cellSpans().get(1).startOffset());
        assertEquals(8, result.cellSpans().get(1).endOffset());
        assertEquals("🎉", slice(source, result.cellSpans().get(0)));
        assertEquals("a", slice(source, result.cellSpans().get(1)));
    }

    @Test
    void threeColumnRow() {
        String source = "| name | age | city |";
        RowCells result = TableCells.parseRowCells(source, 0, source);
        assertEquals(List.of("name", "age", "city"), result.cells());
        assertEquals("name", slice(source, result.cellSpans().get(0)));
        assertEquals("age", slice(source, result.cellSpans().get(1)));
        assertEquals("city", slice(source, result.cellSpans().get(2)));
    }

    @Test
    void delimiterRowGivesDashes() {
        String source = "| --- | --- |";
        RowCells result = TableCells.parseRowCells(source, 0, source);
        assertEquals(List.of("---", "---"), result.cells());
    }

    @Test
    void cellsListIsImmutable() {
        String source = "| a | b |";
        RowCells result = TableCells.parseRowCells(source, 0, source);
        assertTrue(result.cells() instanceof List);
        assertEquals(UnsupportedOperationException.class, unsupportedOnAdd(result));
    }

    private static Class<?> unsupportedOnAdd(RowCells result) {
        try {
            result.cells().add("x");
            return null;
        } catch (UnsupportedOperationException e) {
            return e.getClass();
        }
    }

    private static String slice(String source, Span span) {
        return source.substring(span.startOffset(), span.endOffset());
    }
}

File: rust/core/tests/table_cells_test.rs
//! Port of `TableCellsTest.java` / the table-cell span cases of `table-cells.ts`.
//! The `cellsListIsImmutable` case is dropped (Rust `Vec` is owned/immutable by
//! construction).

use varar_core::offsets::utf16_slice;
use varar_core::span::Span;
use varar_core::table_cells::parse_row_cells;

fn slice(source: &str, span: Span) -> &str {
    utf16_slice(source, span.start_offset, span.end_offset)
}

#[test]
fn basic_row_returns_trimmed_cells() {
    let source = "| a | b |";
    let result = parse_row_cells(source, 0, source);
    assert_eq!(vec!["a".to_string(), "b".to_string()], result.cells);
}

#[test]
fn basic_row_spans_point_to_trimmed_text() {
    let source = "| a | b |";
    let result = parse_row_cells(source, 0, source);
    assert_eq!(2, result.cell_spans.len());
    assert_eq!("a", slice(source, result.cell_spans[0]));
    assert_eq!("b", slice(source, result.cell_spans[1]));
}

#[test]
fn extra_padding_is_trimmed() {
    let source = "| Bob  | 30  |";
    let result = parse_row_cells(source, 0, source);
    assert_eq!(vec!["Bob".to_string(), "30".to_string()], result.cells);
    assert_eq!("Bob", slice(source, result.cell_spans[0]));
    assert_eq!("30", slice(source, result.cell_spans[1]));
}

#[test]
fn no_pipe_returns_empty() {
    let source = "hello world";
    let result = parse_row_cells(source, 0, source);
    assert!(result.cells.is_empty());
    assert!(result.cell_spans.is_empty());
}

#[test]
fn single_pipe_returns_empty() {
    let source = "| only one";
    let result = parse_row_cells(source, 0, source);
    assert!(result.cells.is_empty());
    assert!(result.cell_spans.is_empty());
}

#[test]
fn single_column_table_row() {
    let source = "| n |";
    let result = parse_row_cells(source, 0, source);
    assert_eq!(vec!["n".to_string()], result.cells);
    assert_eq!("n", slice(source, result.cell_spans[0]));
}

#[test]
fn line_start_offset_shifts_spans() {
    let prefix = "# T\n\n";
    let row = "| a | b |";
    let source = format!("{prefix}{row}");
    let line_start = prefix.len(); // ASCII prefix: byte len == utf16 len
    let result = parse_row_cells(row, line_start, &source);
    assert_eq!(vec!["a".to_string(), "b".to_string()], result.cells);
    assert_eq!("a", slice(&source, result.cell_spans[0]));
    assert_eq!("b", slice(&source, result.cell_spans[1]));
}

#[test]
fn astral_cell_shifts_following_span() {
    let source = "| 🎉 | a |"; // U+1F389 PARTY POPPER (2 UTF-16 code units)
    let result = parse_row_cells(source, 0, source);
    assert_eq!(vec!["🎉".to_string(), "a".to_string()], result.cells);
    assert_eq!(2, result.cell_spans[0].start_offset);
    assert_eq!(4, result.cell_spans[0].end_offset);
    assert_eq!(7, result.cell_spans[1].start_offset);
    assert_eq!(8, result.cell_spans[1].end_offset);
    assert_eq!("🎉", slice(source, result.cell_spans[0]));
    assert_eq!("a", slice(source, result.cell_spans[1]));
}

#[test]
fn three_column_row() {
    let source = "| name | age | city |";
    let result = parse_row_cells(source, 0, source);
    assert_eq!(
        vec!["name".to_string(), "age".to_string(), "city".to_string()],
        result.cells
    );
    assert_eq!("name", slice(source, result.cell_spans[0]));
    assert_eq!("age", slice(source, result.cell_spans[1]));
    assert_eq!("city", slice(source, result.cell_spans[2]));
}

#[test]
fn delimiter_row_gives_dashes() {
    let source = "| --- | --- |";
    let result = parse_row_cells(source, 0, source);
    assert_eq!(vec!["---".to_string(), "---".to_string()], result.cells);
}
